services/work-failer/app/workers/failer.ts

```diff
@@ -73,16 +75,30 @@ export default class Failer implements Worker {
   /**
    * Find work items that're older than lastUpdateOlderThanMinutes and call handleWorkItemUpdate.
    * @param lastUpdateOlderThanMinutes - upper limit on the duration since the last update
+   * @param failerDisabledDelay - number of ms to sleep when the failer is disabled because
+   * there are too many items on the work item update queue. Override to a small value in tests.
    * @returns Resolves when the request is complete
    */
-  async handleWorkItemUpdates(lastUpdateOlderThanMinutes: number): Promise<void> {
+  async handleWorkItemUpdates(
+    lastUpdateOlderThanMinutes: number,
+    failerDisabledDelay = 3000,
+  ): Promise<void> {
     let done = false;
     let startingId = 0;
     let numExpired = 0;
     const batchSize = env.workFailerBatchSize;
     log.info('Work failer processing started.');
 
     while (!done) {
+      if (env.maxWorkItemsOnUpdateQueueFailer !== -1) {
+        const smallUpdateQueue = getQueueForType(WorkItemQueueType.SMALL_ITEM_UPDATE);
+        const updateQueueCount = await smallUpdateQueue.getApproximateNumberOfMessages();
+        if (updateQueueCount > env.maxWorkItemsOnUpdateQueueFailer) {
+          log.warn(`Work item update queue is too large with ${updateQueueCount} items, will not fail more work`);
+          await sleep(failerDisabledDelay);
+          continue;
+        }
+      }
       const { workItems, jobServiceThresholds, maxId: newId } = await this.getExpiredWorkItems(lastUpdateOlderThanMinutes, startingId, batchSize);
```

Comment:
I think we should make one more change to set the batchSize to the minimum of `env.workFailerBatchSize` and the number of items that can still be added to the queue.

For example if there are 997 items on the queue with a limit of 1000 and the `env.workFailerBatchSize` is 1000 we should pass in 3 rather than 1000 to the call to `getExpiredWorkItems` so we only fail 3 more things.

Reply:
Updated. Thanks!